common/http/platform/beast/http.cpp

```diff
@@ -353,14 +438,21 @@ io::ExpectedAsyncReadWriterPtr Client::SwitchProtocol(IncomingResponsePtr req) {
 	// This no longer belongs to us.
 	stream_.reset();
 
-	if (is_https_) {
-		return make_shared<RawSocket<ssl::stream<tcp::socket>>>(
+	switch (socket_mode_) {
+	case SocketMode::TlsTls:
+		return make_shared<RawSocket<ssl::stream<ssl::stream<tcp::socket>>>>(
 			stream, response_data_.response_buffer_);
-	} else {
+	case SocketMode::Tls:
+		return make_shared<RawSocket<ssl::stream<tcp::socket>>>(
+			make_shared<ssl::stream<tcp::socket>>(std::move(stream->next_layer())),
+			response_data_.response_buffer_);
+	case SocketMode::Plain:
 		return make_shared<RawSocket<tcp::socket>>(
-			make_shared<tcp::socket>(std::move(stream->next_layer())),
+			make_shared<tcp::socket>(std::move(stream->next_layer().next_layer())),
```

Comment:
Q: `get_lowest_layer()` ?

Reply:
I tried, but this doesn't work, because there is a hidden fourth layer inside `tcp::socket`, `boost::asio::basic_stream_socket`.

Reply:
:see_no_evil: